src/main/java/net/helix/pendulum/service/API.java

```diff
@@ -1498,10 +1504,74 @@ public void publish(BundleTypes type, final String address, final int minWeightM
      * @param txToApprove transactions to approve
      * @throws Exception if storing fails
      */
-    private void storeCustomBundle(final Hash sndAddr, final Hash rcvAddr, List<Hash> txToApprove, byte[] data, final long tag, final int mwm, boolean sign, int keyIdx, int maxKeyIdx, String keyfile, int security) throws Exception {
+    private void storeCustomBundle(final Hash sndAddr, final Hash rcvAddr, List<Hash> txToApprove, byte[] data, final long tag, final int mwm, boolean sign, int keyIdx, int maxKeyIdx, String keyfile, int security, boolean createVirtualTransactions) throws Exception {
         BundleUtils bundle = new BundleUtils(sndAddr, rcvAddr);
         bundle.create(data, tag, sign, keyIdx, maxKeyIdx, keyfile, security);
-        storeAndBroadcast(txToApprove.get(0), txToApprove.get(1), mwm, bundle.getTransactions());
+        List<String> transactionStrings = attachAndStore(txToApprove.get(0), txToApprove.get(1), mwm, bundle.getTransactions());
+
+        if (createVirtualTransactions) {
+            createAndBroadcastVirtualTransactions(data, transactionStrings);
+        }
+        broadcastTransactionsStatement(transactionStrings);
+    }
+
+    /**
+     * Creates and broadcast virtual transactions
+     *
+     * @param tipsBytes
+     * @param milestoneBundle
+     */
+    private void createAndBroadcastVirtualTransactions(byte[] tipsBytes, List<String> milestoneBundle) {
+        TransactionViewModel milestone = getFirstTransactionFromBundle(milestoneBundle);
+        if (milestone != null) {
+            List<Hash> tips = extractTipsFromData(tipsBytes);
```

Comment:
Do we have guarantees here that the tips are solid?

Reply:
No, milestone for current node is not created yet, so tips are not solid (probably the other milestones from other nodes are in the same place).
I thought that we should broadcast virtual transaction before broadcasting the milestone.
Also here all virtual transaction are broadcasted (should be limitted)